rep-2007.rst

```diff
@@ -0,0 +1,252 @@
+
+The primary reason for this change is to improve the developer's experience working with ROS 2.
+Currently, developers often write code to convert a ROS interface into another custom data structure for use in their programs.
+This can be trivial, in the case accessing the ``data`` field in ``std_msgs::msg::String``;
+or more involved such when converting OpenCV's ``cv::Map`` to ROS's ``sensor_msgs/msg/Image`` type [1]_.
+Such conversions are additional work for the programmer and are potential sources of errors.
+
+Terminology
+===========
+
+.. glossary::
+     Custom type (in code, ``CustomType``)
+       A data structure that *is not* a ROS 2 interface, such as ``std::string`` or ``cv::Mat``.
+     ROS type (in code, ``RosType``)
+       A data structure that *is* a ROS 2 interface, such as ``std_msgs::msg::String`` or ``sensor_msgs::msg::Image``.
+
+Specification
+=============
+
+  The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current ROS client libraries, if applicable (roscpp, rospy, roslisp, etc...).
+
+There are different tiers of expressivity in the syntax for type adaptation, including both explicit to implicit forms.
+The explicit forms require the custom type and the ROS type to be specified; 
+while the implicit form makes an assumption about the ROS type based on the custom type.
+The explicit forms may be used to clearly communicate what is happening, which could be useful in examples or tutorials, or may be used depending on the preferences of the developer.
+While the implicit, and most concise, form will primarily be used for its convenience (although it lacks some of the functionality of the explicit forms).
+
+Below are examples of different syntaxes for type adaptation.
+In these examples and using the terminology above, ``std::string`` will be considered the custom type and the ROS interface ``std_msgs::msg::String`` will be considered the ROS type.
+Note that, for every publisher created in the examples below, a ``std::string`` type variable (the custom type) would be adapted to be published on the ROS network as a ``std_msgs::msg::String`` (the ROS type).
+
+.. code-block:: cpp
+
+   // Explicit Form 1: Most explicit
+   using MsgT = rclcpp::AdaptType<std::string>::to<std_msgs::msg::String>;
+   rclcpp::Publisher<MsgT>::SharedPtr publisher1_;
+
+The most explicit form could be shortened in the following to ways:
+
+.. code-block:: cpp
+
+   // Explicit Form 2
+   rclcpp::Publisher<AdaptType<std::string, std_msgs::msg::String>>::SharedPtr publisher2_;
+
+   // Explicit Form 3
+   rclcpp::Publisher<std::string, std_msgs::msg::String>::SharedPtr publisher_3;
```

Comment:
Should this be

```suggestion
   rclcpp::Publisher<std::string, std_msgs::msg::String>::SharedPtr publisher3_;
```

Or should the others be changed? I don't mind either way. Only thing I'd say is sense the context of this code doesn't show them as members in a class, the trailing `_` probably doesn't make sense.

Reply:
I removed the trailing underscores.